File: AlgorithmsProblemSolvingJava/src/io/programminglife/leetcode/Arrays/BestBuySellStock2.java
package io.programminglife.leetcode.Arrays;

/**
 * Say you have an array prices for which the ith element is the price of a given stock on day i.
 * Design an algorithm to find the maximum profit. You may complete as many transactions as you like (i.e., buy one and sell one share of the stock multiple times).
 * Note: You may not engage in multiple transactions at the same time (i.e., you must sell the stock before you buy again).
 * 
 * Input: [7,1,5,3,6,4]
 * Output: 7
 * Explanation: Buy on day 2 (price = 1) and sell on day 3 (price = 5), profit = 5-1 = 4.
 *              Then buy on day 4 (price = 3) and sell on day 5 (price = 6), profit = 6-3 = 3.
 */
public class BestBuySellStock2 {

    public static int maxProfit(int[] prices) {
        int i = 0;
        int valley = prices[0];
        int peak = prices[0];
        int maxProfit = 0;

        while(i < prices.length - 1) {
            while(i < prices.length - 1 && prices[i] >= prices[i + 1]) {
                i++;
            }
            valley = prices[i];
            while(i < prices.length - 1 && prices[i] <= prices[i + 1]) {
                i++;
            }
            peak = prices[i];
            maxProfit += peak - valley;
        }

        return maxProfit;
    }

    public static void main(String[] args) {
        int[] prices = {1,2,3,4,5};

        System.out.println(maxProfit(prices));
    }
    
}

File: AlgorithmsProblemSolvingPython/leetcode/Arrays/bestBuySellStock2.py
def max_profit(prices):
    i = 0
    valley = prices[0]
    peak = prices[0]
    max_profit = 0

    while(i < len(prices) - 1):
        while(i < len(prices) - 1 and prices[i] >= prices[i - 1]):
            i += 1
        valley = prices[i]

        while(i < len(prices) - 1 and prices[i] <= prices[i - 1]):
            i += 1
        peak = prices[i]

        max_profit += peak - valley

    return max_profit

print(max_profit([7,1,5,3,6,4]))